libraries/AP_HAL_PX4/UARTDriver.cpp

```diff
@@ -500,44 +440,33 @@ void PX4UARTDriver::_timer_tick(void)
     _in_timer = true;
 
     // write any pending bytes
-    uint16_t _tail;
-    n = BUF_AVAILABLE(_writebuf);
+    n = _writebuf.available();
     if (n > 0) {
-        uint16_t n1 = _writebuf_size - _writebuf_head;
+        ByteBuffer::IoVec vec[2];
         perf_begin(_perf_uart);
-        if (n1 >= n) {
-            // do as a single write
-            _write_fd(&_writebuf[_writebuf_head], n);
-        } else {
-            // split into two writes
-            int ret = _write_fd(&_writebuf[_writebuf_head], n1);
-            if (ret == n1 && n > n1) {
-                _write_fd(&_writebuf[_writebuf_head], n - n1);                
-            }
+        const auto n_vec = _writebuf.peekiovec(vec, n);
+        for (int i = 0; i < n_vec; i++) {
+            ret = _write_fd(vec[i].data, (uint16_t)vec[i].len);
+            if (ret > 0)
+                _writebuf.advance(ret);
```

Comment:
ret can be positive, but less than the value requested (`vec[i].len`). So if we continue the loop and write to `vec[i + 1].data` we will actually write to the wrong position